python/lsst/butler/config.py

```diff
@@ -0,0 +1,320 @@
+    initialized with a dict. The only caveat is that keys may NOT contain dots ('.'). This is explained next:
+    Config extends the dict api so that hierarchical values may be accessed with dot-delimited notiation.
+    That is, foo.getValue('a.b.c') is the same as foo['a']['b']['c'] is the same as foo['a.b.c'], and either
+    of these syntaxes may be used.
+
+    Storage formats supported:
+    - yaml: read and write is supported.
+    """
+
+    def __init__(self, other=None):
+        """Initialize the Config. Other can be used to initialize the Config in a variety of ways:
+        other (string) Treated as a path to a config file on disk. Must end with '.paf' or '.yaml'.
+        other (Config) Copies the other Config's values into this one.
+        other (dict) Copies the values from the dict into this Config.
+        """
+        collections.UserDict.__init__(self)
+
+        if other is None:
+            return
+
+        if isinstance(other, collections.Mapping):
+            self.update(other)
+        elif isinstance(other, Config):
+            self.data = copy.deepcopy(other.data)
+        elif isinstance(other, str):
+            # if other is a string, assume it is a file path.
+            self.__initFromFile(other)
+        else:
+            # if the config specified by other could not be loaded raise a runtime error.
+            raise RuntimeError("A Config could not be loaded from other:%s" % other)
+
+    def ppprint(self):
+        """helper function for debugging, prints a config out in a readable way in the debugger.
+
+        use: pdb> print myConfigObject.pprint()
+        :return: a prettyprint formatted string representing the config
+        """
+        import pprint
+        return pprint.pformat(self.data, indent=2, width=1)
+
+    def __repr__(self):
+        return self.data.__repr__()
+
+    def __initFromFile(self, path):
+        """Load a file from path. If path is a list, will pick one to use, according to order specified
+        by extensionPreference.
+
+        :param path: string or list of strings, to a persisted config file.
+        :param extensionPreference: the order in which to try to open files. Will use the first one that
+        succeeds.
+        :return:
+        """
+        config = None
+        if path.endswith('yaml'):
+            self.__initFromYamlFile(path)
+        else:
+            raise RuntimeError("Unhandled config file type:%s" % path)
+
+    def __initFromYamlFile(self, path):
+        """Opens a file at a given path and attempts to load it in from yaml.
+
+        :param path:
+        :return:
+        """
+        with open(path, 'r') as f:
+            self.__initFromYaml(f)
+
+    def __initFromYaml(self, stream):
+        """Loads a YAML config from any readable stream that contains one.
+
+        :param stream:
+        :return:
+        """
+        # will raise yaml.YAMLError if there is an error loading the file.
+        self.data = yaml.load(stream)
```

Comment:
Are the files too complex to use `safe_load()`?

Reply:
No idea. Probably. Should check.

Reply:
Apparently not yet. So making it `safe_load` for now.